source/client-side-encryption/tests/README.rst

```diff
@@ -0,0 +1,289 @@
+   - Create a data key with the "aws" KMS provider using `KeyVault.createDataKey()`.
+   - Expect a BSON binary with subtype 4 to be returned.
+   - Expect a `findOne` on the key vault collection with `_id` set to the returned binary to return a document.
+   - Expect that document to have "masterKey.provider" set to "aws"
+
+#. Test explicit encrypt and decrypt with the "local" KMS provider.
+
+   - Create a `KeyVault` with a "local" KMS provider.
+   - Insert the a key document in to the key vault.
+   - Use `KeyVault.encrypt` to encrypt the value "hello" with the following:
+
+     - the algorithm "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
+     - the "local" KMS provider
+
+   - Expect the value is equal to a known BSON binary of subtype 6
+   - Use `KeyVault.decrypt` to decrypt the encrypted value
+   - Expect the value is equal to the string "hello"
+
+#. Test explicit encrypt and decrypt with the "aws" KMS provider.
+
+   - Create a `KeyVault` with a "aws" KMS provider.
+   - Insert the a key document in to the key vault.
+   - Use `KeyVault.encrypt` to encrypt the value "hello" with the following:
+
+     - the algorithm "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
+     - the "aws" KMS provider
+
+   - Expect the value is equal to a known BSON binary of subtype 6
+   - Use `KeyVault.decrypt` to decrypt the encrypted value
+   - Expect the value is equal to the string "hello"
+
+#. Test explicit encrypt of invalid values.
+
+   - Create a `KeyVault` with either a "local" or "aws" KMS provider
+   - Use `KeyVault.encrypt` to attempt to encrypt each BSON type with deterministic encryption.
+
+     - Expect a `string` to succeed. An exception MUST be thrown.
+     - Expect a `document` to fail. An exception MUST be thrown.
+     - Expect a `null` to fail. An exception MUST be thrown.
+     - Expect a BSON binary subtype 6 to fail. An exception MUST be thrown.
+
+   - Use `KeyVault.encrypt` to attempt to encrypt a document using randomized encryption.
+
+     - Expect a `document` to succeed.
+     - Expect a BSON binary subtype 6 to fail. An exception MUST be thrown.
+     - Expect a `null` to fail. An exception MUST be thrown.
+
+#. Test explicit encryption with auto decryption.
+
+   - Create a `KeyVault` with either a "local" or "aws" KMS provider
+   - Use `KeyVault.encrypt` to encrypt a value.
+   - Create a document, setting some field to the value.
+   - Insert the document into a collection.
+   - Find the document. Verify both the value matches the originally set value.
+
+#. Test explicit encrypting an auto encrypted field.
+
+   - Create a `KeyVault` with either a "local" or "aws" KMS provider
+   - Create a collection with a JSON Schema specifying an encrypted field.
+   - Use `KeyVault.encrypt` to encrypt a value.
+   - Create a document, setting the auto-encrypted field to the value.
+   - Insert the document. Verify an exception is thrown.
+
+#. Test bypassing automatic spawning of mongocryptd.
```

Comment:
Does this duplicate the first YAML/JSON test in spawnMongocrypt?

Reply:
It tests the opposite. Since the YAML/JSON test has no way to specify "spawn this process outside of the test", the YAML/JSON test verifies that if you bypass the spawn, the client fails to connect. This tests that starting the mongocryptd process beforehand, and bypassing spawn, succeeds.